src/ThreadDb.cc

```diff
@@ -0,0 +1,232 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "ThreadDb.h"
+#include "GdbServer.h"
+#include "Task.h"
+#include "TaskGroup.h"
+#include "log.h"
+
+extern "C" {
+// The proc_service/thread_db library has a very weird API.  It
+// requires the user of the library to provide certain functions which
+// it links to (rather than, say, having the library user supply a
+// struct of function pointers).  We have to ensure that these
+// functions have C linkage, so that libthread_db can find them.
+#include "proc_service.h"
+}
+
+#include <asm/prctl.h>
+#include <dlfcn.h>
+#include <linux/elf.h>
+#include <sys/reg.h>
+
+#define LIBRARY_NAME "libthread_db.so.1"
+
+// Needed for the logging API.
+using namespace rr;
+
+ps_err_e ps_pglobal_lookup(struct ps_prochandle* h, const char*,
+                           const char* symbol, psaddr_t* sym_addr) {
+  rr::remote_ptr<void> addr;
+  if (!h->db->query_symbol(symbol, &addr)) {
+    LOG(debug) << "ps_pglobal_lookup " << symbol << " failed";
+    return PS_NOSYM;
+  }
+  *sym_addr = reinterpret_cast<psaddr_t>(addr.as_int());
+  LOG(debug) << "ps_pglobal_lookup " << symbol << " OK";
+  return PS_OK;
+}
+
+ps_err_e ps_pdread(struct ps_prochandle* h, psaddr_t addr, void* buffer,
+                   size_t len) {
+  bool ok = true;
+  uintptr_t uaddr = reinterpret_cast<uintptr_t>(addr);
+  // We need any task associated with the task group.  Here we assume
+  // that all the tasks in the task group share VM.
+  rr::Task* task = *h->task_group->task_set().begin();
+  task->read_bytes_helper(uaddr, len, buffer, &ok);
+  LOG(debug) << "ps_pdread " << ok;
+  return ok ? PS_OK : PS_ERR;
+}
+
+ps_err_e ps_pdwrite(struct ps_prochandle*, psaddr_t, const void*, size_t) {
+  LOG(debug) << "ps_pdwrite";
+  return PS_ERR;
+}
+
+ps_err_e ps_lgetregs(struct ps_prochandle* h, lwpid_t tid, prgregset_t result) {
+  rr::Task* task = h->task_group->session()->find_task_tid(tid);
+  assert(task != nullptr);
+
+  struct ::user_regs_struct regs = task->regs().get_ptrace();
+  memcpy(result, static_cast<void*>(&regs), sizeof(regs));
+  LOG(debug) << "ps_lgetregs OK";
+  return PS_OK;
+}
+
+ps_err_e ps_lsetregs(struct ps_prochandle*, lwpid_t, const prgregset_t) {
+  LOG(debug) << "ps_lsetregs";
+  return PS_ERR;
+}
+
+ps_err_e ps_lgetfpregs(struct ps_prochandle*, lwpid_t, prfpregset_t*) {
+  LOG(debug) << "ps_lgetfpregs";
+  return PS_ERR;
+}
+
+ps_err_e ps_lsetfpregs(struct ps_prochandle*, lwpid_t, const prfpregset_t*) {
+  LOG(debug) << "ps_lsetfpregs";
+  return PS_ERR;
+}
+
+pid_t ps_getpid(struct ps_prochandle* h) {
+  LOG(debug) << "ps_getpid " << h->task_group->tgid;
+  return h->task_group->tgid;
```

Comment:
libthread_db only calls ps_getpid an unusual situation, I think maybe when it's being used before NPTL has fully initialized.